core/dbt/parser/schemas.py

```diff
@@ -561,39 +624,71 @@ def parse_exposures(self, block: YamlBlock) -> None:
 
     def parse_file(self, block: FileBlock) -> None:
         dct = self._yaml_from_file(block.file)
-        # mark the file as seen, even if there are no macros in it
+
+        # mark the file as seen, in ParseResult.files
         self.results.get_file(block.file)
+
         if dct:
             try:
+                # This does a deep_map to check for circular references
                 dct = self.raw_renderer.render_data(dct)
             except CompilationException as exc:
                 raise CompilationException(
                     f'Failed to render {block.path.original_file_path} from '
                     f'project {self.project.project_name}: {exc}'
                 ) from exc
 
+            # contains the FileBlock and the data (dictionary)
             yaml_block = YamlBlock.from_file_block(block, dct)
 
-            self._parse_format_version(yaml_block)
+            # checks version
+            self._check_format_version(yaml_block)
 
             parser: YamlDocsReader
-            for key in NodeType.documentable():
-                plural = key.pluralize()
-                if key == NodeType.Source:
-                    parser = SourceParser(self, yaml_block, plural)
-                elif key == NodeType.Macro:
-                    parser = MacroPatchParser(self, yaml_block, plural)
-                elif key == NodeType.Analysis:
-                    parser = AnalysisPatchParser(self, yaml_block, plural)
-                elif key == NodeType.Exposure:
-                    # handle exposures separately, but they are
-                    # technically still "documentable"
-                    continue
-                else:
-                    parser = TestablePatchParser(self, yaml_block, plural)
+
+            # There are 6 kinds of parsers:
+            # Model, Seed, Snapshot, Source, Macro, Analysis,
+
+            # NonSourceParser.parse(), TestablePatchParser is a variety of
+            # NodePatchParser
+            if 'models' in dct:
```

Comment:
ah yes, this is a lot easier to grok

Reply:
Yes, easier to understand what's happening. Also a tiny bit of a performance improvement because we're not constructing a bunch of parsers and then throwing them away when we check to see if there are any entries for that key...